### PR TITLE
fix(prometheus): implement graceful shutdown

### PR DESCRIPTION
**Pull request by zamazan4ik**
- implement graceful shutdown if a port cannot be taken on the start

Tested:
- Local run

### CODE REVIEW

**Comment by zamazan4ik** [comment]
Partially fixes https://github.com/vectordotdev/vector/issues/15527

File: src/sinks/prometheus/exporter.rs [hunk 1]
```diff
@@ -515,13 +511,16 @@ impl PrometheusExporter {
...
     }
 }
 
 #[async_trait]
 impl StreamSink<Event> for PrometheusExporter {
     async fn run(mut self: Box<Self>, mut input: BoxStream<'_, Event>) -> Result<(), ()> {
-        self.start_server_if_needed().await;
+        self.start_server_if_needed()
+            .await
+            .map_err(|error| error!("Start server error: {}.", error))?;
```
Comment (tobz) [inline src/sinks/prometheus/exporter.rs:523]: This error message should be something more appropriate, like `Failed to start Prometheus exporter: {}.`.

**Review by tobz — CHANGES_REQUESTED** [review]

File: src/sinks/prometheus/exporter.rs (hunk 1 above)
Comment (zamazan4ik) [reply src/sinks/prometheus/exporter.rs]: Fixed

**Review by bruceg — APPROVED** [review]
LGTM

**Comment by bruceg** [comment]
I'm not seeing why this has almost a 20% reported performance regression on the `http_pipelines_blackhole_acks` test. This code path is not even used for that, is it?

**Comment by tobz** [comment]
> I'm not seeing why this has almost a 20% reported performance regression on the `http_pipelines_blackhole_acks` test. This code path is not even used for that, is it?

The `http_pipelines_*` soaks have been flaky as all hell recently: I swear I've seen them go from 15% to -20% and back and forth between different PRs, etc. I was discussing the possibility of just disabling them entirely because `pipelines` is a hidden/not even really used transform, so the soaks for it mostly burn compute for something that we effectively don't care about. I had also talked to SMP about theses like these and they had thoughts on trying to better detect tests that need more replicas, etc, to get more reliable results.

Tactically, though, I think we can safely ignore them here.

**Review by davidhuie-dd — APPROVED** [review]

**Comment by jszwedko** [comment]
Thanks for this contribution @zamazan4ik !

**Pull request merged**